KAFKA-6246: Dynamic update of listeners and security configs

rajinisivaram opened the pull request:
Dynamic update of listeners as described in KIP-226. This includes:
  - Addition of new listeners with listener-prefixed security configs
  - Removal of existing listeners
  - Password encryption
  - sasl.jaas.config property for broker's JAAS config prefixed with listener and mechanism name

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

rajinisivaram committed:
KAFKA-6246: Dynamic update of listeners and security configs

rajinisivaram commented:
@hachikuji This is currently a commit on top of the thread pool resize PR. I will rebase once that is committed. Can you review, please? Thank you!

rajinisivaram commented:
@hachikuji I have rebased this after merging the other two. Thanks.

rajinisivaram commented:
retest this please

hachikuji commented on clients/src/main/java/org/apache/kafka/common/Reconfigurable.java:
```diff
@@ -33,9 +35,12 @@
-     * if this method returns false or throws any exception.
+     * if this method throws any exception.
+     * @throws ConfigException if the provided configs are not valid. The exception
+     *         message from ConfigException will be returned to the client in
```
nit: in _the_ AlterConfigs response

hachikuji commented on clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java:
```diff
@@ -204,6 +204,15 @@ public Password getPassword(String key) {
+     * <ul>
+     *     <li>listener.name.{listenerName}.some.prop: If the provided prefix is `listener.name.{listenerName}.`,
+     *         the key `some.prop` with the value parsed using the definition of `some.prop` is returned.</li>
+     *     <li>listener.name.{listenerName}.anotherprefix.some.prop: If the provided prefix is `listener.name.{listenerName}.`,
```
It would be helpful here to mention the specific use case for this extended prefix option.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/security/JaasContext.java:
```diff
@@ -56,42 +56,44 @@
      */
     public static JaasContext load(JaasContext.Type contextType, ListenerName listenerName,
-                                   Map<String, ?> configs) {
+                                   Map<String, ?> configs, String mechanism) {
```
It is a little curious that the mechanism is unused for clients. If we don't need it, maybe we should have separate functions for the client and server. The client also has no need for `listenerName` apparently.

rajinisivaram replied:
Separated out the methods for client/server context type.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/240/

hachikuji commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1209,6 +1227,29 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
     }
   }
 
+  case class BrokerModifications(brokerId: Int) extends ControllerEvent {
```
We don't appear to use `brokerId`.

hachikuji commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1209,6 +1227,29 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
   }
 
+  case class BrokerModifications(brokerId: Int) extends ControllerEvent {
+    override def state: ControllerState = ControllerState.TopicChange
```
Should this be `BrokerChange` or a new state?

rajinisivaram replied:
Updated to `BrokerChange`

hachikuji commented on clients/src/main/java/org/apache/kafka/common/network/SaslChannelBuilder.java:
```diff
@@ -110,8 +114,14 @@ public void configure(Map<String, ?> configs) throws KafkaException {
-            this.loginManager = LoginManager.acquireLoginManager(jaasContext, hasKerberos, configs);
-
+            for (Map.Entry<String, JaasContext> entry : jaasContexts.entrySet()) {
+                String mechanism = entry.getKey();
```
It seems now the mechanisms are passed in the keys of this map, so does that mean we can remove the `clientSaslMechanism` argument and infer `hasKerberos` above from the map?

rajinisivaram replied:
I think we need `clientSaslMechanism` since for servers, there can be multiple entries in the map and one `clientSaslMechanism`. I have simplified `hasKerberos`, but again I think we need it for brokers.

hachikuji replied:
I am probably missing something, but it seems like we pass null as the `clientMechanism` for servers. For clients, we use a singleton map with the entry key equal to the `clientMechanism`, so that's why I thought it was redundant.

rajinisivaram replied:
Clients have a singleton jaas config map. Brokers making a server connection may have multiple entries in the jaas config map and `clientMechanism` is null. Brokers making a client connection for inter-broker communication may have multiple entries in the jaas config map and a non-null client mechanism.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/security/JaasContext.java:
```diff
@@ -56,42 +56,44 @@
-            }
+            JaasConfig jaasConfig = new JaasConfig(globalContextName, jaasConfigArgs.value());
+            AppConfigurationEntry[] contextModules = jaasConfig.getAppConfigurationEntry(globalContextName);
+            int numModules = contextModules == null ? 0 : contextModules.length;
```
Not part of this patch, but since we're modifying this code anyway... It's super weird to assign 0 to this variable and then immediately check for that case and raise an exception. Wouldn't it be more direct to throw the exception if `contextModules` is null? The second check could also just be based on `contextModules.length`.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/security/authenticator/CredentialCache.java:
```diff
@@ -16,13 +16,12 @@
-    private final Map<String, Cache<? extends Object>> cacheMap = new HashMap<>();
+    private final Map<String, Cache<? extends Object>> cacheMap = new ConcurrentHashMap<>();
 
     public <C> Cache<C> createCache(String mechanism, Class<C> credentialClass) {
```
If the concurrent map is important, shouldn't we rewrite this to use `putIfAbsent`?

rajinisivaram replied:
We don't really need this to be a ConcurrentHashMap since we create all the caches on broker startup, but changed to `putIfAbsent` just in case we change that in future.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslServerAuthenticator.java:
```diff
@@ -154,6 +152,12 @@ public SaslServerAuthenticator(Map<String, ?> configs,
             throw new IllegalArgumentException("No SASL mechanisms are enabled");
         this.enabledMechanisms = new HashSet<>(enabledMechanisms);
+        for (String mechanism : enabledMechanisms) {
+            if (jaasContexts.get(mechanism) == null)
```
Should this be `containsKey` or is a `null` a possible value we need to check for? Same question for `subjects`.

rajinisivaram replied:
Changed to `containsKey`

hachikuji commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -187,7 +189,18 @@ class SocketServer(val config: KafkaConfig, val metrics: Metrics, val time: Time
+
+  def removeListeners(listenersRemoved: Seq[EndPoint]): Unit = {
+    listenersRemoved.foreach { endpoint =>
+      acceptors.remove(endpoint).foreach(_.shutdown())
```
Do we need the lock to remove from `acceptors`?

rajinisivaram replied:
Added locking and changed to ConcurrentHashMap since it is read without a lock.

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -219,17 +242,50 @@ class DynamicBrokerConfig(private val kafkaConfig: KafkaConfig) extends Logging
+  private def maybeReEncodePasswords(persistentProps: Properties, adminZkClient: AdminZkClient): Properties = {
+    val props = persistentProps.clone().asInstanceOf[Properties]
+    if (!props.asScala.keySet.exists(DynamicPasswordConfigs.contains)) {
+      maybeCreatePasswordEncoder(kafkaConfig.passwordEncoderOldSecret).foreach { passwordDecoder =>
```
By the way, we should make a note to add some documentation on dynamic broker configs and key changes.

rajinisivaram replied:
Added a note to KAFKA-6476

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -357,11 +417,18 @@ class DynamicBrokerConfig(private val kafkaConfig: KafkaConfig) extends Logging
+          if (needsReconfiguration(reconfigurable.reconfigurableConfigs.asJava, updatedMap.keySet)) {
+            if (validateOnly)
+              reconfigurable.validateReconfiguration(newConfig)
+            else
```
Aren't we skipping validation before updating the current config?

rajinisivaram replied:
Oops, fixed.

hachikuji replied:
I'm looking for the fix, but it seems like we still update the current config without validating when `validateOnly` is false.

rajinisivaram replied:
Sorry, fixed another case, missed this one. Updated.

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -548,9 +618,10 @@ class DynamicMetricsReporters(brokerId: Int, server: KafkaServer) extends Reconf
-        !updatedMetricsReporters.contains(reporter.getClass.getName) || reporter.validateReconfiguration(configs)
+        if (updatedMetricsReporters.contains(reporter.getClass.getName))
+          reporter.validateReconfiguration(configs)
       case _ => true
```
Seems we could leave this empty

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -588,3 +659,103 @@ class DynamicMetricsReporters(brokerId: Int, server: KafkaServer) extends Reconf
+    val newListeners = listenersToMap(newConfig.listeners)
+    val newAdvertisedListeners = listenersToMap(newConfig.advertisedListeners)
+    val oldListeners = listenersToMap(oldConfig.listeners)
+    if (newListeners.keySet != newAdvertisedListeners.keySet)
```
Do we need equality? I thought the advertised listeners could be a subset of the listeners.

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -588,3 +659,103 @@ class DynamicMetricsReporters(brokerId: Int, server: KafkaServer) extends Reconf
+      if (changed)
+        throw new ConfigException(s"Existing listener configs cannot be updated dynamically, restart broker or create a new listener for update")
+    }
+    if (!newListeners.contains(newConfig.interBrokerListenerName))
```
We also need a check somewhere that the inter-broker listener is among the new advertised listeners.

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -588,3 +659,103 @@ class DynamicMetricsReporters(brokerId: Int, server: KafkaServer) extends Reconf
+      val prefix = listenerName.configPrefix
+      val newListenerProps = immutableListenerConfigs(newConfig, prefix)
+      val oldListenerProps = immutableListenerConfigs(oldConfig, prefix)
+      val changed = newListenerProps.size != oldListenerProps.size || newListenerProps.exists { case (k, v) =>
```
Is this the same thing as checking `newListenerProps != oldListenerProps`?

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -588,3 +659,103 @@ class DynamicMetricsReporters(brokerId: Int, server: KafkaServer) extends Reconf
+          !oldListenerProps.contains(k) || oldListenerProps(k) != v
+      }
+      if (changed)
+        throw new ConfigException(s"Existing listener configs cannot be updated dynamically, restart broker or create a new listener for update")
```
Perhaps we should mention the specific listener name in this message?

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -385,8 +452,12 @@ class DynamicBrokerConfig(private val kafkaConfig: KafkaConfig) extends Logging
+        reconfigurable.validateReconfiguration(newConfigs)
+      } catch {
+        case e: ConfigException => throw e
+        case _: Exception => throw new ConfigException("Validation of dynamic config update failed with class " + reconfigurable.getClass)
```
Should we mention the configs that we tried to update?

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -385,8 +452,12 @@ class DynamicBrokerConfig(private val kafkaConfig: KafkaConfig) extends Logging
+        case _: Exception => throw new ConfigException("Validation of dynamic config update failed with class " + reconfigurable.getClass)
+      }
     } else
       reconfigurable.reconfigure(newConfigs)
```
Maybe we could have an info log message which mentioned the reconfigurable and the custom configs that are being updated?

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -588,3 +659,103 @@ class DynamicMetricsReporters(brokerId: Int, server: KafkaServer) extends Reconf
+    val newListenerMap = listenersToMap(newListeners)
+    val oldListeners = oldConfig.listeners
+    val oldListenerMap = listenersToMap(oldListeners)
+    val listenersRemoved = oldListeners.filterNot(e => newListenerMap.contains(e.listenerName))
```
Just so I understand, is it possible to remove a listener which has been configured statically?

rajinisivaram replied:
Yes, any listener can be removed apart from the current inter-broker listener.

hachikuji commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -236,8 +237,9 @@ class KafkaServer(val config: KafkaConfig, time: Time = Time.SYSTEM, threadNameP
-        tokenCache = new DelegationTokenCache(config.saslEnabledMechanisms)
-        credentialProvider = new CredentialProvider(config.saslEnabledMechanisms, tokenCache)
+        // Enable delegation token cache for all SCRAM mechanisms to simplify dynamic update
+        tokenCache = new DelegationTokenCache(ScramMechanism.mechanismNames)
```
I'm not sure I understand why it makes sense to refer to the scram mechanisms here in place of `saslEnabledMechanisms`.

rajinisivaram replied:
We watch ZooKeeper for changes anyway, so rather than create the cache only if a mechanism is enabled, the code currently updates the cache anyway. Delegation tokens uses SCRAM, hence it is for all SCRAM mechanisms.

hachikuji commented on core/src/main/scala/kafka/utils/PasswordEncoder.scala:
```diff
@@ -0,0 +1,160 @@
+  val PasswordLengthProp = "passwordLength"
+}
+
+class PasswordEncoder(secret: Password,
```
I am wondering if this should be implemented in the common package in case we ever have need to reuse on the client side.

rajinisivaram replied:
At the moment, it is tailored for password encryption rather than generic encryption. If we need this functionality on the client-side, we can decide whether to reuse this or not. For now, shall I just leave it in core?

hachikuji replied:
Yes, sounds fine.

hachikuji commented on core/src/main/scala/kafka/utils/PasswordEncoder.scala:
```diff
@@ -0,0 +1,160 @@
+}
+
+class PasswordEncoder(secret: Password,
+                      keyFactoryAlgorithm: Option[String],
```
Might be helpful to document somewhere that these parameters are only used in the encoding process. The decoding process depends on obtaining them from the encoded value.

hachikuji reviewed (commented):
Thanks for the patch. I am still reviewing the test cases, but thought I may as well post the comments I have.

rajinisivaram committed:
Address review comments

hachikuji commented on clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java:
```diff
@@ -204,6 +204,16 @@ public Password getPassword(String key) {
+     * </ul>
+     * </p>
      */
     public Map<String, Object> valuesWithPrefixOverride(String prefix) {
```
Can you update the test cases for this function to include the new usage?

rajinisivaram commented:
@hachikuji Thanks for the review, have addressed the comments.

hachikuji commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1227,8 +1227,8 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
 
-  case class BrokerModifications(brokerId: Int) extends ControllerEvent {
-    override def state: ControllerState = ControllerState.TopicChange
+  case class BrokerModifications() extends ControllerEvent {
```
nit: this could be a `case object`

hachikuji commented on clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java:
```diff
@@ -144,12 +144,11 @@ else if (clientAuthConfig.equals("requested"))
                 createSSLContext(newKeystore);
-            return true;
         } catch (Exception e) {
             throw new KafkaException("Validation of dynamic config update failed", e);
```
Should we throw `ConfigException`?

hachikuji commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1209,6 +1227,29 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
+          case None => false
+        }
+      }
+      val updatedBrokerIdsSorted = updatedBrokers.map(_.id).toSeq.sorted
```
Maybe we can move this into the `if` below and use check `updatedBrokers.nonEmpty` instead? It avoids the log message when there are no updated brokers.

hachikuji commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1209,6 +1227,29 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
     }
   }
 
+  case class BrokerModifications() extends ControllerEvent {
```
Use `case object`?

hachikuji commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -164,9 +164,9 @@ class SocketServer(val config: KafkaConfig, val metrics: Metrics, val time: Time
   def resizeThreadPool(oldNumNetworkThreads: Int, newNumNetworkThreads: Int): Unit = {
```
Would it be safer to synchronize this as well?

hachikuji commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1458,7 +1499,17 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
+class BrokerModificationsHandler(controller: KafkaController, eventManager: ControllerEventManager, brokerId: Int) extends ZNodeChangeHandler {
+  override val path: String = BrokerIdZNode.path(brokerId)
+
+  override def handleDataChange(): Unit = {
```
Perhaps we should have a log message?

rajinisivaram replied:
Since we have a log message in the process() method of BrokerModifications, do we need another one here? For other events, logging is in `process()` method of the event.

hachikuji commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -185,9 +185,20 @@ class SocketServer(val config: KafkaConfig, val metrics: Metrics, val time: Time
+    }
+  }
+
+  def addListeners(listenersAdded: Seq[EndPoint]): Unit = synchronized {
```
Add log line her and in `removeListeners`?

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -588,3 +653,106 @@ class DynamicMetricsReporters(brokerId: Int, server: KafkaServer) extends Reconf
+    val oldListeners = listenersToMap(oldConfig.listeners)
+    val oldAdvertisedListeners = listenersToMap(oldConfig.advertisedListeners)
+    if (!newAdvertisedListeners.keySet.subsetOf(newListeners.keySet))
+      throw new ConfigException(s"Listeners '$newListeners' and advertisedListeners '$newAdvertisedListeners' don't match")
```
The log message needs to be updated.

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -588,3 +653,106 @@ class DynamicMetricsReporters(brokerId: Int, server: KafkaServer) extends Reconf
+        throw new ConfigException(s"Configs cannot be updated dynamically for existing listener $listenerName, " +
+          "restart broker or create a new listener for update")
+    }
+    if (!newListeners.contains(newConfig.interBrokerListenerName))
```
nit: this check is not necessary since we have the check below and we know `newAdvertisedListeners` is a subset of `newListeners`.

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -588,3 +653,106 @@ class DynamicMetricsReporters(brokerId: Int, server: KafkaServer) extends Reconf
+      throw new ConfigException(s"Cannot delete inter-broker listener ${newConfig.interBrokerListenerName}")
+    if (!newAdvertisedListeners.contains(newConfig.interBrokerListenerName))
+      throw new ConfigException(s"Advertised listener must be specified for inter-broker listener ${newConfig.interBrokerListenerName}")
+    else if (!oldAdvertisedListeners.contains(newConfig.interBrokerListenerName))
```
If I understand this correctly, we are allowing the inter-broker listener to change, but only to one of the other listeners that are currently advertised. Is that right? Also, the message below could be clarified a little since it doesn't actually mention the inter-broker listener property.

rajinisivaram replied:
I had forgotten that we don't allow changing inter-broker listener at all. I had moved it out to `Future Work` alongside message format version in the KIP. Removed the check.

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -704,6 +719,16 @@ object KafkaConfig {
+  /** ********* Password encryption configuration for dynamic configs *********/
+  val PasswordEncoderSecretDoc = "The secret used for encoding dynamically configured passwords for this broker."
+  val PasswordEncoderOldSecretDoc = "The old secret that was used for encoding dynamically configured passwords. " +
+    "This is required only when secret is updated. If specified, all dynamically encoded passwords are " +
```
nit: only when _the_ secret is updated.

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -704,6 +719,16 @@ object KafkaConfig {
+  val PasswordEncoderOldSecretDoc = "The old secret that was used for encoding dynamically configured passwords. " +
+    "This is required only when secret is updated. If specified, all dynamically encoded passwords are " +
+    s"decoded using this old secret and re-encoded using $PasswordEncoderSecretProp when broker starts up."
+  val PasswordEncoderKeyFactoryAlgorithmDoc = "The SecretKeyFactory algorithm used for encoding dynamically configured passwords." +
```
nit: space at the end

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -704,6 +719,16 @@ object KafkaConfig {
+  val PasswordEncoderKeyFactoryAlgorithmDoc = "The SecretKeyFactory algorithm used for encoding dynamically configured passwords." +
+    "Default is PBKDF2WithHmacSHA512 if available and PBKDF2WithHmacSHA1 otherwise."
+  val PasswordEncoderCipherAlgorithmDoc = "The Cipher algorithm used for encoding dynamically configured passwords."
+  val PasswordEncoderKeyLengthDoc =  s"The key length used for encoding dynamically configured passwords."
```
nit: drop the "s" since there are no substitutions.

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -911,6 +937,13 @@ object KafkaConfig {
+      /** ********* Password encryption configuration for dynamic configs *********/
+      .define(PasswordEncoderSecretProp, PASSWORD, null, MEDIUM, PasswordEncoderSecretDoc)
+      .define(PasswordEncoderOldSecretProp, PASSWORD, null, MEDIUM, PasswordEncoderOldSecretDoc)
+      .define(PasswordEncoderKeyFactoryAlgorithmProp, STRING, null, MEDIUM, PasswordEncoderKeyFactoryAlgorithmDoc)
```
I wonder if the importance could be LOW for this and the configs below since the defaults seem pretty reasonable?

hachikuji commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -236,8 +237,10 @@ class KafkaServer(val config: KafkaConfig, time: Time = Time.SYSTEM, threadNameP
-        tokenCache = new DelegationTokenCache(config.saslEnabledMechanisms)
-        credentialProvider = new CredentialProvider(config.saslEnabledMechanisms, tokenCache)
+        // Enable delegation token cache for all SCRAM mechanisms to simplify dynamic update.
+        // This keeps the cache uptodate if new SCRAM mechanisms are enabled dynamically.
```
nit: up-to-date

hachikuji commented on core/src/main/scala/kafka/utils/PasswordEncoder.scala:
```diff
@@ -0,0 +1,173 @@
+}
+
+/**
+  * Password encoder and decoder implementation.
```
May also be worth documenting the encoding format

hachikuji commented on core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala:
```diff
@@ -532,13 +538,203 @@ class DynamicBrokerReconfigurationTest extends ZooKeeperTestHarness with SaslSet
   }
 
+  @Test
+  def testAdvertisedListenerUpdate(): Unit = {
```
Is it worthwhile having a test case for changing the inter-broker-listener?

rajinisivaram replied:
Sorry for the confusion about inter-broker listeners. Added a test to verify that inter-broker listener cannot be updated.

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -588,3 +653,106 @@ class DynamicMetricsReporters(brokerId: Int, server: KafkaServer) extends Reconf
+      throw new ConfigException(s"Listeners '$newListeners' and advertisedListeners '$newAdvertisedListeners' don't match")
+    if (newListeners.keySet != newConfig.listenerSecurityProtocolMap.keySet)
+      throw new ConfigException(s"Listeners '$newListeners' and listener map '${newConfig.listenerSecurityProtocolMap}' don't match")
+    newListeners.keySet.intersect(oldListeners.keySet).foreach { listenerName =>
```
Does this also cover the case in which the security mapping of an existing listener is changed?

rajinisivaram replied:
Missed that one, added.

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -223,6 +223,11 @@ object Defaults {
+  val PasswordEncoderIterations = 4096
 }
 
 object KafkaConfig {
```
When we update documentation, maybe we can add an additional column to the configuration table which indicates whether the property is dynamically configurable.

hachikuji commented on core/src/test/scala/unit/kafka/server/DynamicBrokerConfigTest.scala:
```diff
@@ -151,4 +152,65 @@ class DynamicBrokerConfigTest {
+  }
+
+  @Test
+  def testPasswordConfigEncoderSecretChange(): Unit = {
```
I assumed this covered the reencoding the passwords after a secret change, but it looks a little different. Should we cover that case too?

hachikuji commented on core/src/test/scala/unit/kafka/utils/PasswordEncoderTest.scala:
```diff
@@ -0,0 +1,129 @@
+  }
+
+
+}
```
nit: newline

hachikuji reviewed (commented):
Thanks for the updates. Left a few more comments.

rajinisivaram committed:
Address review comments

rajinisivaram commented:
@hachikuji Thanks for the review (and for merging the other PR). I have addressed the comments and rebased.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/297/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/298/

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -424,8 +424,7 @@ class DynamicBrokerConfig(private val kafkaConfig: KafkaConfig) extends Logging
           if (needsReconfiguration(reconfigurable.reconfigurableConfigs.asJava, updatedMap.keySet)) {
             if (validateOnly)
```
Feels like I'm being difficult, but shouldn't the logic be like this?

```scala
             reconfigurable.validateReconfiguration(newConfig)
             if (!validateOnly)
               brokerReconfigurablesToUpdate += reconfigurable
```

rajinisivaram replied:
Sorry, I am running out of excuses! Updated and added unit test.

rajinisivaram committed:
Address review comments

rajinisivaram commented:
@hachikuji Thanks for the review. Addressed the comment and removed unused code.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/300/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/301/

hachikuji reviewed (approved):
Thanks for the updates. LGTM. Can you take a look at the failing test cases? I think one of them is among the new tests in this patch.

rajinisivaram committed:
Fix transient test failure

rajinisivaram commented:
@hachikuji Thanks for the review. There was a timing issue with `DynamicBrokerReconfigurationTest#testAdvertisedListenerUpdate()`. I have pushed a fix for that. Don't think the other test failures are related to the PR.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/311/

rajinisivaram committed:
Fix TestUtils.deleteBrokersInZk

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/312/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/313/

rajinisivaram commented:
@hachikuji Can you take a look at the last two commits before I merge? Thank you!

The first is the fix for the transient test failure in `DynamicBrokerReconfigurationTest#testAdvertisedListenerUpdate()`. I have run the tests locally 100 times and it passed and the builds have passed too.

I thought the other test failure was unrelated, but it turns out it was caused by `EndPoint#toString` added in this PR. I reverted that change since the default `toString` is good enough and we have been using it in other log entries anyway. The test failure was actually due to an existing bug in `TestUtils.deleteBrokersInZk` which didn't cause an issue before because the old toString produced a valid (but wrong) ZK path, but the new one produced an invalid ZK path. Fixed `TestUtils.deleteBrokersInZk` as well.

hachikuji commented:
LGTM, thanks for fixing the tests. I will merge to trunk and 1.1.

junrao commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1209,6 +1239,29 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
+    override def process(): Unit = {
+      if (!isActive) return
+      val curBrokers = zkClient.getAllBrokersInCluster.toSet
+      val updatedBrokers = controllerContext.liveBrokers.filter { broker =>
```
@rajinisivaram : Sorry for the late review. Just a minor comment. It seems that we can avoid reading all brokers from ZK by passing in brokerId to BrokerModifications and only read the info for that brokerId?

rajinisivaram replied:
@junrao Thanks for the review. Opened PR #4670.
